### PR TITLE
Companion PR for for indexer's `tzaffi/box-ingest`

### PR DESCRIPTION
**Pull request by tzaffi**
## Summary

* Adding `LookupKv()` to `indexerLedgerForEval interface` (this is implemented in Indexer on `type LedgerForEvaluator`)
* Implemeting `LookupKv()` on:
  * `indexerLedgerConnector`
  * `accountUpdatesLedgerEvaluator`
* Adding a new `logic` function `SplitBoxKey()` which is basically the inverse of `MakeBoxKey()`

## Companion PR

Indexer's [#1067](https://github.com/algorand/indexer/pull/1067) which ingests boxes into indexer's DB.

## Test Plan

* Added unit test `data/transactions/logic/box_test.go::TestMakeBoxKey()`
* Manual test of network that creates 125 apps with 1000 boxes and verify directly on IndexerDB that app_box is populated as expected
* Unit and Integration tests exist in the Indexer PR

### CODE REVIEW

File: data/transactions/logic/box.go
Comment (tzaffi) [inline data/transactions/logic/box.go:180]: This is useful for box ingestion in indexer, but in theory, it could live there, not here. Open to suggestions.

File: network/wsNetwork.go
Comment (tzaffi) [inline network/wsNetwork.go:733]:
```suggestion
	wn.broadcastQueueBulk = make(chan broadcastRequest, 10)
```

oops, shouldn't have sent upstream

File: ledger/evalindexer_test.go
Comment (tzaffi) [inline ledger/evalindexer_test.go:111]: TODO: add more tests here

File: logging/log.go
Comment (tzaffi) [inline logging/log.go:353]: This diff will go away when master is merged back into `feature/avm-box`

File: ledger/acctupdates.go
Comment (tzaffi) [inline ledger/acctupdates.go:323]: TODO: remove comment before merging.

File: ledger/acctupdates.go
Comment (tzaffi) [inline ledger/acctupdates.go:847]: TODO: remove comments before merging

File: data/transactions/logic/box.go
Comment (tzaffi) [inline data/transactions/logic/box.go:191]: TODO ^^^

File: data/transactions/logic/box.go
Comment (tzaffi) [inline data/transactions/logic/box.go:159]: introducing constants to be shared between `MakeBoxKey()` and `GetAppAndNameFromKey()`

File: ledger/evalindexer.go
Comment (tzaffi) [inline ledger/evalindexer.go:46]: `indexerLedgerConnector.LookupKv()` gains access to `*ledger.Ledger.LookupKv()` via its field `il: indexerLedgerForEval`.

File: ledger/acctupdates.go
Comment (tzaffi) [inline ledger/acctupdates.go:846]: Appreciate feedback here ^^^

File: ledger/acctupdates.go
Comment (tzaffi) [inline ledger/acctupdates.go:845]: Need to implement as migration from sqlite uses `Ledger.accts: accountUpdates` when importing blocks: cf. indexer's [MakeLedger](https://github.com/algorand/indexer/blob/5584c6dca2f2f3a492f7650aeb905ab24ffac682/util/ledger_util.go#L72) opens the ledger of type `Ledger`

File: data/transactions/logic/box.go
Comment (tzaffi) [inline data/transactions/logic/box.go:188]:
```suggestion
	return app, string(keyBytes[boxNameIndex:]), nil
```

File: data/transactions/logic/box.go
Comment (tzaffi) [inline data/transactions/logic/box.go:185]: To gain a sliver of speed, am dispensing with the check that `key[:boxPrefixLength] == boxPrefix`

File: data/transactions/logic/box.go
Comment (jannotti) [reply data/transactions/logic/box.go:180]: I'd like something simpler like `SplitBoxKey` but not intending to bike shed

File: data/transactions/logic/box.go
Comment (jannotti) [inline data/transactions/logic/box.go:181]: Since you're using an `error` return value, why not do the prefix check?

File: data/transactions/logic/box.go
Comment (jannotti) [inline data/transactions/logic/box.go:187]: I had no idea this cast syntax `(basics.AppIndex)x` was legal Go. It seems much more common to use `basics.AppIndex(x)`

File: data/transactions/logic/box.go
Comment (jannotti) [inline data/transactions/logic/box.go:188]:
```suggestion
	return app, key[boxNameIndex:], nil
```

File: data/transactions/logic/box.go
Comment (jannotti) [inline data/transactions/logic/box.go:182]:
```suggestion
	if len(key) < boxNameIndex {
```

File: ledger/acctupdates.go
Comment (jannotti) [reply ledger/acctupdates.go:846]: I doubt it.  How do the other lookup routines in `accountUpdatesLedgerEvaluator` work?  There should be ones for resources and accounts.

Comment (tzaffi) [reply ledger/acctupdates.go:846]: The two [functions above](https://github.com/algorand/go-algorand/blob/0632f1f9ee599f5f865cb9ee0bf7a0fabc21e7d0/ledger/acctupdates.go#L835-L843) (`LookupApplication` and `LookupAsset`) don't take a lock. Similarly, neither does `GetCreatorForRound()` nor `LookupWithoutRewards()`.

The [2 year old comment on L779](https://github.com/algorand/go-algorand/blob/0632f1f9ee599f5f865cb9ee0bf7a0fabc21e7d0/ledger/acctupdates.go#L779) can be interpreted as implying that the purpose of `accountUpdatesLedgerEvaluator` is to allow not taking the lock.

File: data/transactions/logic/box.go
Comment (tzaffi) [inline data/transactions/logic/box.go:182]:
```suggestion
		return 0, "", fmt.Errorf("SplitBoxKey() cannot extract AppIndex as key (%s) too short (length=%d)", key, len(key))
```

File: data/transactions/logic/box.go
Comment (tzaffi) [inline data/transactions/logic/box.go:185]:
```suggestion
		return 0, "", fmt.Errorf("SplitBoxKey() cannot extract AppIndex as key (%s) too short (length=%d)", key, len(key))
	}
	if key[:boxPrefixLength] != boxPrefix {
		return 0, "", fmt.Errorf("SplitBoxKey() illegal app box prefix in key (%s). Expected prefix '%s'", key, boxPrefix)
```

File: ledger/acctupdates.go
Comment (jannotti) [reply ledger/acctupdates.go:845]: This gets to my question above about avoiding this cache loading jazz for the indexer.  This code will eventually be needed anyway, so the only bad thing that may come from you doing it with indexer work is that we have to get this code merged into feature branch sooner than your entire indexer work is done.  Nothing wrong with that, I suppose.

File: data/transactions/logic/box.go
Comment (shiqizng) [inline data/transactions/logic/box.go:189]: what if len(key)!=boxNameIndex?

Comment (tzaffi) [reply data/transactions/logic/box.go:189]: The case `len(key) < boxNameIndex` is handled above which leaves only the case `len(key) > boxNameIndex` to address your question. That case is perfectly fine, since `key[boxNameIndex:]` stores the box name

File: ledger/acctupdates.go
Comment (tzaffi) [inline ledger/acctupdates.go:973]:
```suggestion
```

File: ledger/evalindexer.go
Comment (tzaffi) [inline ledger/evalindexer.go:152]:
```suggestion
	value, err := l.il.LookupKv(rnd, key)
	if err != nil {
```

**Review by tzaffi — COMMENTED** [review]
changes

File: .circleci/config.yml
Comment (jannotti) [inline .circleci/config.yml:419]: But should be merged to feature branch?

File: Makefile
Comment (jannotti) [inline Makefile:259]: I feel I should NOT MERGE THIS CHANGE.

File: test/scripts/e2e.sh
Comment (jannotti) [inline test/scripts/e2e.sh:179]: if seems indented with spaces, but `fi` with tab?

File: test/scripts/e2e.sh
Comment (jannotti) [inline test/scripts/e2e.sh:205]: Seems excessive?

42?

File: test/scripts/e2e_client_runner.py
Comment (jannotti) [inline test/scripts/e2e_client_runner.py:474]: Now that there's an `else`, I'd rather swap the predicate so it's not testing `not`.

File: .circleci/config.yml
Comment (tzaffi) [reply .circleci/config.yml:419]: no, they should not. I'll open a new branch shortly that contains the experimental changes.

File: Makefile
Comment (tzaffi) [reply Makefile:259]: correct

File: test/scripts/e2e.sh
Comment (tzaffi) [reply test/scripts/e2e.sh:205]: all experimental and should have been in a seperate branch.

**Review by jannotti — APPROVED** [review]

**Pull request merged**
